### DISCUSSION TITLE
Curious about creating an S2CellID from a FaceIJ and level

### DISCUSSION THREAD

**Discussion by CraigglesO** [google/s2geometry#221] [General]
I've been looking at the code for a while and curious if there is a creative way to go from FaceIJ+level to S2CellID and back again

To expound on this:

if The max level was say "2", then the maximum for I and J are "4".
The idea here is that I have an algorithm that sometimes uses IJ at various depths and I'm curious the easiest way to get to an S2CellID that complements the depth.

should I just bit shift the IJ to maxLevel, create an S2CellID then find its parent at the level I was originally in or do you have a superior alternative?

essentially I assume it is something like this before FromIJ:
```js
I = I << (30 - level)
J = J << (30 - level)
```

then after getting the s2cell parent at level, heading back and after ToIJ:
```js
i = i >> (30 - level)
j = j >> (30 - level)
```

while the math works from tests, I'm nervous there is something I am missing. The way I visualize it though, this seems to work.

**Comment by jmr** [comment]
You're looking for something like `S2CellId::FromFaceIJ().parent()`? Can you use that?

**Reply by CraigglesO** [reply]
Hey, sorry if I was not being clear enough. I'll try again.

FromFaceIJ then parent just ensures its a S2CellId as a specific level. But what I am interested in is working with IJ at various levels.

But I think the best way to explain this is to use different variables. Let's say X and Y.

So at level 0 we only have a face or just a singular cell. But at the next level (1) we have 2 cells in the X direction and 2 cells in the Y direction with 4 cells total.

My original fear was that after converting to IJ, I couldn't get the correct XY from IJ. but after more experimentation I know I can get the correct "XY" from "IJ" for sure now.

it really is just:
```cpp
int IJtoXY (int i, int level) {
  return i >> (30 - level);
}
```

**Reply by CraigglesO** [reply]
However, I think it's interesting that from the main page [here](https://s2geometry.io/resources/earthcube) the top right is considered 2-2 whereas my IJtoXY would categorize the top right as 3-3.

**Reply by jmr** [reply]
I'm still not exactly sure what you're after. Do you want to "work with" (i, j), or convert to and from cell ids?

Did you read the coordinate descriptions here?

https://github.com/google/s2geometry/blob/master/src/s2/s2coords.h#L30

Maybe you want a shifted version of `(si, ti)`? You can do that.

**Reply by CraigglesO** [reply]
Let's put it this way. If it was you, looking at the page I linked in the prior comment, second image down.. how would YOU arrive at 2-2? It's clear it is using level 2 cells, but given a S2CellID, how would you arrive at that 2-2 number? for the top right cell at level/zoom 2.

**Comment by jmr** [comment]
>  looking at the page I linked in the prior comment, second image down.. how would YOU arrive at 2-2?

Using Python pseudo-code: `[cell_id.child_position(level) for level in range(1, cell_id.level())]`

This is equivalent to reading off pairs of bits from the cell id:

https://github.com/google/s2geometry/blob/ad1489e898f369ca09e2099353ccd55bd0fd7a26/src/s2/s2cell_id.h#L66

https://github.com/google/s2geometry/blob/ad1489e898f369ca09e2099353ccd55bd0fd7a26/src/s2/s2cell_id.h#L599

But if that's your question, I'm confused as to why you mention `(i, j)` above.

**Reply by CraigglesO** [reply]
Hey, that was a huge insight for me. I thought it was specifying an x-y but ti was actually just explaining consecutive cell positions in the image.
wow.

Ok, so I made some pictures to explain what I am doing:

<img width="791" alt="Screen Shot 2021-12-13 at 16 04 55" src="https://user-images.githubusercontent.com/5249140/145890029-b66922f5-acc1-46f2-b7d0-0c10128eab5b.png">

<img width="791" alt="Screen Shot 2021-12-13 at 16 05 11" src="https://user-images.githubusercontent.com/5249140/145890053-9779aa2c-b654-48d9-b539-69b16e1ffee3.png">

<img width="791" alt="Screen Shot 2021-12-13 at 16 23 56" src="https://user-images.githubusercontent.com/5249140/145891105-8a2382da-2e86-4b82-929c-a7e17456359c.png">


I got a bit lazy and didn't fill it all in, but you can see what is going on. Sometimes I am interested in the "X-Y" pair positions from a S2CellID.

In the second image, the top right position has an XY of `[3, 3]`. However if I converted that S2CellID to IJ I would actually get `[469762048, 469762048]`.

The reason is, no matter the S2CellID `level`, the I and J values will always be between `[0, pow(2, 30) - 1]`.

I still believe my function is the correct/best method to get from an S2CellID to an XY at that cells level
if the level is 30, than XY and IJ will be the same, otherwise it "dives"

```cpp
int IJtoXY (int i, int level) {
  return i >> (30 - level);
}
```

So:
1) convert S2CellID to a Face-I-J
2) convert the I and J to X and Y

Is this a bit more clear for what I want? I don't think s2geometry provides this function.

**Comment by jmr** [comment]
Yes, that will work. There's probably no function for that since it's just a shift. 

What are you actually trying to do? Unless you're writing a tutorial about the coordinate systems, this is probably using the library at a lower level than necessary.

**Reply by CraigglesO** [reply]
Sorry, was a bit sick.

Javascript is a unique situation where I can:
1) Use "long" package and add a ton of storage overhead.
2) Use BigInt and see a 99% reduction in "number" performance.
3) Use Face-zoom(level)-x-y and manipulate data as numbers, then convert to BigInt as necessary.

3 has the nice advantage of speed and no package/size cost. But basically I was here to verify that the x-y solution was actually accurate, but in hindsight it was pretty obviously yes.

I appreciate your time though, this is really an absurdly impressive library.
